ONSdigital/eq-author-app
# Fixes to ID issues that was causing problems with EAR-1485 and EAR-1397

### Context of PR ###
Fixes issues with duplicating ID's causing problems with question deletion, question duplication, section duplication as well as moving questions between folders when one has been deleted. 

Before the remap is done we create a new page using the existing page as a reference, this is ok for the root attributes, but nested attributes are not cloned, but copied by ref. To fix this, we `cloneDeep` meaning all the nested attributes are cloned. We're doing this in most place, but we weren't doing it on the `createQuestionPage` resolver.

When i added `inPlace` to the `remapAllNestedIds` function, meaning `deepMap` didn't create a new object, but changed the existing object, since it was referencing nested attributes and not cloning them it changed the ID's of the original object meaning the ID's were the same.

### How to review ###

- Create a page with an answer.
- Duplicate that page.
- Delete the duplicated answer and check it doesn't delete the original answer.

- Create a page with a Date Answer.
- Add an Advanced Property.
- Set a Max or Min Value.
- Turn on Max or Min Value.
- Add any number of days with a `custom` date.
- Duplicate that page and check it doesn't break, if it does break you'll get an error page.
- Duplicate that section and check it doesn't break, if it does break you'll get an error page.

- Create multiple folders with multiple pages within each folder.
- Delete a question inside of a folder.
- Check that you can still move pages between folders.

### What to do after everything is green ###

    *  Bring this branch up-to-date with Origin/Master
    *  If there are a lot of commits or some are not helpful to read, squash them down
    *  Click the Merge button on this pull request
    *  Does this change mean the Capability examples questionnaire should be updated?
    *  Move the Jira ticket for this task into the next stage of the process
